extensions-core/s3-extensions/src/main/java/org/apache/druid/storage/s3/LazyObjectSummariesIterator.java

```diff
@@ -0,0 +1,162 @@
+
+    prepareNextRequest();
+    fetchNextBatch();
+    advanceObjectSummary();
+  }
+
+  private void prepareNextRequest()
+  {
+    final URI currentUri = prefixesIterator.next();
+    final String currentBucket = currentUri.getAuthority();
+    final String currentPrefix = S3Utils.extractS3Key(currentUri);
+
+    request = new ListObjectsV2Request()
+        .withBucketName(currentBucket)
+        .withPrefix(currentPrefix)
+        .withMaxKeys(maxListingLength);
+  }
+
+  private void fetchNextBatch()
+  {
+    try {
+      result = S3Utils.retryS3Operation(() -> s3Client.listObjectsV2(request));
+      request.setContinuationToken(result.getNextContinuationToken());
+      objectSummaryIterator = result.getObjectSummaries().iterator();
+    }
+    catch (Exception e) {
+      throw new RE(
+          e,
+          "Failed to get object summaries from S3 bucket[%s], prefix[%s]",
+          request.getBucketName(),
+          request.getPrefix()
+      );
+    }
+  }
+
+  /**
+   * Advance objectSummaryIterator to the next non-placeholder, updating "currentObjectSummary".
+   */
+  private void advanceObjectSummary()
+  {
+    while (objectSummaryIterator.hasNext() || result.isTruncated() || prefixesIterator.hasNext()) {
+      while (objectSummaryIterator.hasNext()) {
+        currentObjectSummary = objectSummaryIterator.next();
+
+        if (!isDirectoryPlaceholder(currentObjectSummary)) {
+          return;
+        }
+      }
+
+      // Exhausted "objectSummaryIterator" without finding a non-placeholder.
+      if (result.isTruncated()) {
+        fetchNextBatch();
+      } else if (prefixesIterator.hasNext()) {
+        prepareNextRequest();
+        fetchNextBatch();
+      }
+    }
+
+    // Truly nothing left to read.
+    currentObjectSummary = null;
+  }
+
+  @Override
+  public boolean hasNext()
+  {
+    return currentObjectSummary != null;
+  }
+
+  @Override
+  public S3ObjectSummary next()
+  {
+    if (currentObjectSummary == null) {
+      throw new NoSuchElementException();
+    }
+
+    final S3ObjectSummary retVal = currentObjectSummary;
+    advanceObjectSummary();
+    return retVal;
+  }
+
+  /**
+   * Checks if a given object is a directory placeholder and should be ignored.
+   *
+   * Adapted from org.jets3t.service.model.StorageObject.isDirectoryPlaceholder(). Does not include the check for
+   * legacy JetS3t directory placeholder objects, since it is based on content-type, which isn't available in an
+   * S3ObjectSummary.
+   */
+  private static boolean isDirectoryPlaceholder(final S3ObjectSummary objectSummary)
+  {
+    // Recognize "standard" directory place-holder indications used by Amazon's AWS Console and Panic's Transmit.
+    if (objectSummary.getKey().endsWith("/") && objectSummary.getSize() == 0) {
+      return true;
+    }
+
+    // Recognize s3sync.rb directory placeholders by MD5/ETag value.
+    if ("d66759af42f282e1ba19144df2d405d0".equals(objectSummary.getETag())) {
```

Comment:
It's from `org.jets3t.service.model.StorageObject.isDirectoryPlaceholder` (mentioned in the javadoc for this method). Sources are at https://github.com/mondain/jets3t/blob/master/jets3t/src/main/java/org/jets3t/service/model/StorageObject.java.

I did it this way for these two reasons:

- Minimal diff from the original function, in case someone wants to compare them visually and see what we changed.
- I didn't see a benefit to putting `"d66759af42f282e1ba19144df2d405d0"` into a constant: it won't change, it won't be used in any other locations, and it's adequately explained by the comment above the line it appears in, so the usual benefits of pulling magic values out into constants don't apply.

Although actually, maybe this is moot, since it might be reasonable to just remove this block. I don't think we have any particular need to support the `s3sync.rb` style directory placeholders. The other two kinds are much more common. It looks like `s3sync.rb` hasn't been maintained in many years, according to http://s3sync.net/wiki.html.

What do you think?

Reply:
Wow I feel embarrassed for not reading that comment 😶

Given that we're close to a release, I'd err on the side of caution and just leave this as you have it. I don't see this creating a maintenance burden, so the benefits for removing it is minimal.

Reply:
In that case, I'll leave it as is. Thanks for taking a look.